### PR TITLE
new method to manage resource in engine

### PR DESCRIPTION
**Pull request by LHT129**
issue: #203

### CODE REVIEW

File: include/vsag/engine.h [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+#include "index.h"
+#include "resource.h"
+
+namespace vsag {
+class Engine {
+public:
+    explicit Engine(Resource* resource = nullptr);
+
+    void
+    Stop();
```
Comment (jiaweizone) [inline include/vsag/engine.h]: Shutdown ?

**Review by wxyucs — COMMENTED** [review]
@LHT129 please add an example of this change

File: src/engine.cpp [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+            CHECK_ARGUMENT(parsed_params.contains(INDEX_DISKANN),
+                           fmt::format("parameters must contains {}", INDEX_DISKANN));
+            auto& diskann_param_obj = parsed_params[INDEX_DISKANN];
+            auto diskann_params =
+                DiskannParameters::FromJson(diskann_param_obj, index_common_params);
+            logger::debug("created a diskann index");
+            return std::make_shared<DiskANN>(diskann_params, index_common_params);
+        } else if (name == INDEX_HGRAPH) {
+            if (allocator == nullptr) {
+                index_common_params.allocator_ = DefaultAllocator::Instance().get();
```
Comment (jiaweizone) [inline src/engine.cpp:83]: minor: `allocator` will definitely not be empty

File: src/resource_owner_wrapper.h [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+
+    ~ResourceOwnerWrapper() override {
+        if (owner_) {
+            delete resource_;
+        }
+    }
+
+private:
+    Resource* resource_{nullptr};
+    bool owner_{false};
```
Comment (jiaweizone) [inline src/resource_owner_wrapper.h]: owner_ -> owned_

File: src/safe_allocator.h [hunk 4]
```diff
@@ -68,6 +69,8 @@ class SafeAllocator : public Allocator {
 private:
     Allocator* const raw_allocator_ = nullptr;
     std::shared_ptr<Allocator> owned_allocator_ = nullptr;
+
+    bool owner_{false};
```
Comment (jiaweizone) [inline src/safe_allocator.h:73]: need deallocator to manage raw_allocator ?

**Comment by LHT129** [comment]
> @LHT129 please add an example of this change, and also need create an issue about this feature

issue is necessary, but I will introduce several pull requests for this feature, this one is the first, the following is the example, test and remove the origin factory method

File: src/resource_owner_wrapper.h (hunk 3 above)
Comment (LHT129) [reply src/resource_owner_wrapper.h]: done

File: include/vsag/engine.h (hunk 1 above)
Comment (LHT129) [reply include/vsag/engine.h]: done

File: src/engine.cpp (hunk 2 above)
Comment (LHT129) [reply src/engine.cpp:83]: the allocator of engine may have many problems, I'd like to fix these in an issue with several PRs. this PR is the first to show the basic structure of the engine without example and test

File: src/safe_allocator.h (hunk 4 above)
Comment (LHT129) [reply src/safe_allocator.h:73]: ditto

File: src/engine.cpp (hunk 2 above)
Comment (jiaweizone) [reply src/engine.cpp:83]: OK

**Review by jiaweizone — APPROVED** [review]
LGTM

**Review by wxyucs — APPROVED** [review]
lgtm

**Pull request merged**